rust/hyperlane-base/src/server/server.rs

```diff
@@ -0,0 +1,113 @@
+use axum::{
+    http::{Response, StatusCode},
+    routing::get,
+    Router,
+};
+use hyper::Body;
+use prometheus::{Encoder, Registry};
+use std::{net::SocketAddr, sync::Arc};
+use tokio::task::JoinHandle;
+use tracing::warn;
+
+/// A server that serves agent-specific routes
+pub struct Server {
+    listen_port: u16,
+    registry: Registry,
+}
+
+impl Server {
+    /// Create a new server instance.
+    pub fn new(listen_port: u16, registry: Registry) -> Self {
+        Self {
+            listen_port,
+            registry,
+        }
+    }
+
+    /// Run an HTTP server serving agent-specific different routes
+    ///
+    /// routes:
+    ///   - metrics - serving OpenMetrics format reports on `/metrics`
+    ///     (this is compatible with Prometheus, which ought to be configured to scrape this endpoint)
+    pub fn run(self: Arc<Self>) -> JoinHandle<()> {
+        let port = self.listen_port;
+        tracing::info!(port, "starting prometheus server on 0.0.0.0");
+
+        let server_clone = self.clone();
+        tokio::spawn(async move {
+            let app = Router::new().route(
+                "/metrics",
+                get(move || {
+                    let server = server_clone.clone();
+                    async move {
+                        match server.gather() {
+                            Ok(metrics) => Response::builder()
+                                // OpenMetrics specs demands "application/openmetrics-text;
+                                // version=1.0.0; charset=utf-8"
+                                // but the prometheus scraper itself doesn't seem to care?
+                                // try text/plain to make web browsers happy.
+                                .header("Content-Type", "text/plain; charset=utf-8")
+                                .body(Body::from(metrics))
+                                .unwrap(),
+                            Err(_) => Response::builder()
+                                .status(StatusCode::NOT_FOUND)
+                                .body(Body::from("Failed to encode metrics"))
+                                .unwrap(),
+                        }
+                    }
+                }),
+            );
+
+            let addr = SocketAddr::from(([0, 0, 0, 0], port));
+            axum::Server::bind(&addr)
+                .serve(app.into_make_service())
+                .await
+                .expect("Failed to start server");
+            warn!("Prometheus server could not be started or exited early");
```

Comment:
I believe this warning would always be printed, regardless of success or failure